src/Formik.tsx

```diff
@@ -1,661 +1,847 @@
-      this.resetForm(this.props.initialValues);
-    }
-  }
-
-  setErrors = (errors: FormikErrors<Values>) => {
-    this.setState({ errors });
-  };
-
-  setTouched = (touched: FormikTouched<Values>) => {
-    this.setState({ touched }, () => {
-      if (this.props.validateOnBlur) {
-        this.runValidations(this.state.values);
-      }
-    });
-  };
-
-  setValues = (values: FormikState<Values>['values']) => {
-    this.setState({ values }, () => {
-      if (this.props.validateOnChange) {
-        this.runValidations(values);
-      }
-    });
-  };
+/** @private */
+function usePrevious<T>(val: T) {
+  const ref = React.useRef<T>(val);
+  React.useEffect(() => {
+    ref.current = val;
+  }, [val]);
+  return ref.current;
+}
 
-  setStatus = (status?: any) => {
-    this.setState({ status });
-  };
+export function useFormik<Values = object>({
+  validateOnChange = true,
+  validateOnBlur = true,
+  isInitialValid,
+  enableReinitialize = false,
+  onSubmit,
+  ...rest
+}: FormikConfig<Values>) {
+  const props = { validateOnChange, validateOnBlur, onSubmit, ...rest };
+  const initialValues = React.useRef(props.initialValues);
+  const initialErrors = React.useRef(props.initialErrors || {});
+  const initialTouched = React.useRef(props.initialTouched || {});
+  const initialStatus = React.useRef(props.initialStatus);
+  const isMounted = React.useRef<boolean>(false);
+  const fields = React.useRef<{
+    [field: string]: {
+      validate: (value: any) => string | Promise<string> | undefined;
+    };
+  }>({});
 
-  setError = (error: any) => {
+  React.useEffect(() => {
     if (process.env.NODE_ENV !== 'production') {
-      console.warn(
-        `Warning: Formik\'s setError(error) is deprecated and may be removed in future releases. Please use Formik\'s setStatus(status) instead. It works identically. For more info see https://github.com/jaredpalmer/formik#setstatus-status-any--void`
+      invariant(
+        typeof isInitialValid === 'undefined',
+        'isInitialValid has been deprecated and will be removed in future versions of Formik. Please use initialErrors instead.'
       );
     }
-    this.setState({ error });
-  };
+  }, [isInitialValid]);
 
-  setSubmitting = (isSubmitting: boolean) => {
-    if (this.didMount) {
-      this.setState({ isSubmitting });
-    }
-  };
+  React.useEffect(() => {
+    isMounted.current = true;
+
+    return () => {
+      isMounted.current = false;
+    };
+  }, []);
+
+  const [state, dispatch] = React.useReducer<
+    React.Reducer<FormikState<Values>, FormikMessage<Values>>
+  >(formikReducer, {
+    values: props.initialValues,
+    errors: props.initialErrors || {},
+    touched: props.initialTouched || {},
+    status: props.initialStatus,
+    isSubmitting: false,
+    isValidating: false,
+    submitCount: 0,
+  });
+  const prevState = usePrevious(state);
+
+  const runValidateHandler = React.useCallback(
+    (values: Values, field?: string): Promise<FormikErrors<Values>> => {
```

Comment:
This isn't particular to this method, but I do have a question about how these callbacks are implemented. And it's a big one (sorry).

I think that `useFormik`'s sole job is basically to gather the hooks, get the data, and hook up the context. But the majority of this function is actually taken up with defining other functions. In addition, the testability of these functions is now non-existent. Finally, it's nearly impossible to determine the control flow of the function itself without collapsing every function bit by bit. So I propose one general strategy with two possible methods:

Strategy: define helpers elsewhere. It could be in this file outside of the function, or a helpers file.

```
/** @private */
function runValidations(validate, values, field) { /* the function */ }

export function useFormik<Values = object>() {
    // etc etc
    // Strategy 1: Use bind (please don't hurt me). 
    const runValidateHandler = React.useCallback(runValidations.bind(null, props.validate), [props.validate]);
    // Strategy 2: Connect the callbacks manually. 
    // Unfortunately, I don't think you can infer values and field like I've done below, 
    // so this would require defining types twice.
    const runValidateHandler = React.useCallback((values, field) => {
        runValidations(props.validate, values, field);
    }, [props.validate]);
}
```

Now, you can unit test `runValidations` really easily without trying to "unit test React". I don't know if there is a way to spy on the function when using `bind` to use `expect(runValidations).toHaveBeenCalled()` or whatever, but if so that's even better.

The best part of this method is when you read this file, you'll just see the following, you can _read_ whether or not everything is hooked up in the correct order and let TypeScript do the rest, and finally a quick glance will allow you to compare that all of the props passed to the helper function are contained in the dependency array.

```
const state = useState(etc, etc);
const initialValues = useRef<>();
const runValidationsHandler = useCallback(callback, [state]);
const runValidationSchema = useCallback(callback, [initialValues]);
const allHandlers = [
    runValidationsHandler,
    runValidationSchema,
];

Reply:
Testing these methods internally is not a goal. These are implementation details. There is a valid usage of `useFormik`. For folks that want to use the state only version. That's it. I think we should likely rename this function to something that is less obvious or mark it as `unstable_`

Reply:
@jaredpalmer I agree with you for the most part. I re-read through the code and for the most part what we are doing in these callbacks is dealing with promises / dispatches.

However, there are some parts I think should be broken out and unit tested. I'll itemize those in a different issue.